pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/ControllerLeaderLocatorIntegrationTest.java

```diff
@@ -0,0 +1,171 @@
+    TestUtils.waitForCondition(aVoid -> {
+      resultSet.clear();
+      for (Map.Entry<Integer, String> entry : _partitionToTableMap.entrySet()) {
+        // Mock the behavior that 40 seconds have passed.
+        controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
+        controllerLeaderLocator.invalidateCachedControllerLeader();
+
+        String tableName = entry.getValue();
+        Pair<String, Integer> pair1 = controllerLeaderLocator.getControllerLeader(tableName);
+        if (pair1 == null) {
+          return false;
+        }
+        resultSet.add(pair1.getFirst() + pair1.getSecond());
+      }
+      // There are two combinations in the result set.
+      return resultSet.size() == 2;
+    }, TIMEOUT_IN_MS, "Failed to get two pairs of controllers.");
+
+    // Disable lead controller resource
+    enableResourceConfigForLeadControllerResource(false);
+
+    TestUtils.waitForCondition(aVoid -> {
+      resultSet.clear();
+      for (Map.Entry<Integer, String> entry : _partitionToTableMap.entrySet()) {
+        // Mock the behavior that 40 seconds have passed.
+        controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
+        controllerLeaderLocator.invalidateCachedControllerLeader();
+
+        String tableName = entry.getValue();
+        Pair<String, Integer> pair1 = controllerLeaderLocator.getControllerLeader(tableName);
+        if (pair1 == null) {
+          return false;
+        }
+        resultSet.add(pair1.getFirst() + pair1.getSecond());
+      }
+      // Only 1 controller should be fetched, which is the helix leader.
+      return resultSet.size() == 1;
+    }, TIMEOUT_IN_MS, "Failed to get only one pair of controller");
+
+    // Mock the behavior that 40 seconds have passed.
+    controllerLeaderLocator.setLastCacheInvalidateMillis(System.currentTimeMillis() - 40_000L);
+    controllerLeaderLocator.invalidateCachedControllerLeader();
+
+    // Since lead controller resource disabled, helix leader should be used.
+    Pair<String, Integer> thirdPair = controllerLeaderLocator.getControllerLeader(testTableName1);
+
+    Assert.assertEquals(pair.getFirst(), thirdPair.getFirst());
+    Assert.assertEquals(pair.getSecond(), thirdPair.getSecond());
+
+    // Stop the second controller.
+    secondControllerStarter.stop();
+  }
+
+  /**
+   * Find the table names for all the partitions.
+   */
+  private void findTableNamesForAllPartitions() {
+    _partitionToTableMap = new HashMap<>();
+
+    int count = 0;
+    while (_partitionToTableMap.size() < Helix.NUMBER_OF_PARTITIONS_IN_LEAD_CONTROLLER_RESOURCE) {
+      String tableName = "testTable" + count;
+      int partitionId = LeadControllerUtils.getPartitionIdForTable(tableName);
+      _partitionToTableMap.putIfAbsent(partitionId, tableName);
+      count++;
+    }
+    System.out.println("Successfully generate partition to table map after " + count + " iterations");
```

Comment:
remove this line, it should not change between test runs, right?

Reply:
Removed.